src/cryptsetup/cryptsetup.c

```diff
@@ -565,11 +584,96 @@ static char *friendly_disk_name(const char *src, const char *vol) {
         return name_buffer;
 }
 
+static PasswordRegisteredType check_registered_passwords(struct crypt_device *cd) {
+
+        int slots = 0, slot_max;
+        PasswordRegisteredType r = _PASSWORD_REGISTERED_TYPE_INVALID;
+
+        assert(cd);
+
+        if (!streq_ptr(crypt_get_type(cd), CRYPT_LUKS2)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Non-LUKS2 devices only support passphrase.");
+                return PASSWORD_PASSPHRASE;
+        }
+
+        /* Search all used slots */
+        assert_se((slot_max = crypt_keyslot_max(CRYPT_LUKS2)) > 0);
+        for (int slot = 0; slot < slot_max; slot++) {
+                crypt_keyslot_info status;
+
+                status = crypt_keyslot_status(cd, slot);
+                if (!IN_SET(status, CRYPT_SLOT_ACTIVE, CRYPT_SLOT_ACTIVE_LAST))
+                        continue;
+
+                slots += slot + 1;
```

Comment:
Hmm, I'm not sure if this works. If slots can be referenced multiple times, subtraction below could arrive at 0 even if some slot is not referenced. Can you explain (in a comment) how this is supposed to work?
(It might be easier to just use a set or a `bool array[slot_max]`).

Reply:
I think you're right, and currently for LUKS2 it's at most 32 key-slots, so better keep the logic simple.